File: 173.二叉搜索树迭代器.cpp
/*
 * @lc app=leetcode.cn id=173 lang=cpp
 *
 * [173] 二叉搜索树迭代器
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class BSTIterator
{
private:
    vector<int> binaryTree;

public:
    void middleScan(TreeNode *root)
    {
        if (!root)
            return;
        middleScan(root->left);
        binaryTree.push_back(root->val);
        middleScan(root->right);
    }
    BSTIterator(TreeNode *root)
    {
        middleScan(root);
    }

    /** @return the next smallest number */
    int next()
    {
        int result = binaryTree[0];
        binaryTree.erase(binaryTree.begin());
        return result;
    }

    /** @return whether we have a next smallest number */
    bool hasNext()
    {
        return binaryTree.size();
    }
};

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator* obj = new BSTIterator(root);
 * int param_1 = obj->next();
 * bool param_2 = obj->hasNext();
 */
// @lc code=end

File: 173.二叉搜索树迭代器.py
#
# @lc app=leetcode.cn id=173 lang=python3
#
# [173] 二叉搜索树迭代器
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class BSTIterator:

    def __init__(self, root: TreeNode):
        self.binaryTree = []

        def middleScan(root: TreeNode):
            if not root:
                return
            middleScan(root.left)
            self.binaryTree.append(root.val)
            middleScan(root.right)

        middleScan(root)

    def next(self) -> int:
        """
        @return the next smallest number
        """
        return self.binaryTree.pop(0)

    def hasNext(self) -> bool:
        """
        @return whether we have a next smallest number
        """
        return bool(len(self.binaryTree))


# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()
# @lc code=end
